File: test/ModularPipelines.UnitTests/Dependencies/DependsOnModulesInCategoryAttributeTests.cs
using ModularPipelines.Attributes;
using ModularPipelines.TestHelpers;

namespace ModularPipelines.UnitTests.Dependencies;

/// <summary>
/// Unit tests for <see cref="DependsOnModulesInCategoryAttribute"/>.
/// </summary>
public class DependsOnModulesInCategoryAttributeTests : TestBase
{
    [Test]
    public async Task ShouldDependOn_ModuleInCategory_ReturnsTrue()
    {
        var attr = new DependsOnModulesInCategoryAttribute("infrastructure");
        var context = new MockDependencyContext().WithCategory(typeof(TestModule), "infrastructure");

        var result = attr.ShouldDependOn(typeof(TestModule), context);

        await Assert.That(result).IsTrue();
    }

    [Test]
    public async Task ShouldDependOn_ModuleInDifferentCategory_ReturnsFalse()
    {
        var attr = new DependsOnModulesInCategoryAttribute("infrastructure");
        var context = new MockDependencyContext().WithCategory(typeof(TestModule), "build");

        var result = attr.ShouldDependOn(typeof(TestModule), context);

        await Assert.That(result).IsFalse();
    }

    [Test]
    public async Task ShouldDependOn_CaseInsensitive_ReturnsTrue()
    {
        var attr = new DependsOnModulesInCategoryAttribute("INFRASTRUCTURE");
        var context = new MockDependencyContext().WithCategory(typeof(TestModule), "infrastructure");

        var result = attr.ShouldDependOn(typeof(TestModule), context);

        await Assert.That(result).IsTrue();
    }

    [Test]
    public async Task ShouldDependOn_ModuleHasNoCategory_ReturnsFalse()
    {
        var attr = new DependsOnModulesInCategoryAttribute("infrastructure");
        var context = new MockDependencyContext();

        var result = attr.ShouldDependOn(typeof(TestModule), context);

        await Assert.That(result).IsFalse();
    }

    [Test]
    public async Task ShouldDependOn_CategoryMatchesExactly_ReturnsTrue()
    {
        var attr = new DependsOnModulesInCategoryAttribute("build-pipeline");
        var context = new MockDependencyContext().WithCategory(typeof(TestModule), "build-pipeline");

        var result = attr.ShouldDependOn(typeof(TestModule), context);

        await Assert.That(result).IsTrue();
    }

    [Test]
    public async Task Constructor_WithNullCategory_ThrowsArgumentException()
    {
        await Assert.That(() => new DependsOnModulesInCategoryAttribute(null!))
            .ThrowsException();
    }

    [Test]
    public async Task Constructor_WithEmptyCategory_ThrowsArgumentException()
    {
        await Assert.That(() => new DependsOnModulesInCategoryAttribute(string.Empty))
            .ThrowsException();
    }

    [Test]
    public async Task Constructor_WithWhitespaceCategory_ThrowsArgumentException()
    {
        await Assert.That(() => new DependsOnModulesInCategoryAttribute("   "))
            .ThrowsException();
    }

    [Test]
    public async Task Category_Property_ReturnsConstructorValue()
    {
        var attr = new DependsOnModulesInCategoryAttribute("my-category");

        await Assert.That(attr.Category).IsEqualTo("my-category");
    }

    private class TestModule { }
}

File: test/ModularPipelines.UnitTests.FSharp/Dependencies/DependsOnModulesInCategoryAttributeTests.fs
namespace ModularPipelines.UnitTests.FSharp.Dependencies

open System
open ModularPipelines.Attributes
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Assertions.FSharp.Operations
open TUnit.Core

type CategoryAttributeTestModule() =
    class
    end

type DependsOnModulesInCategoryAttributeTests() =
    [<Test>]
    member _.ShouldDependOn_ModuleInCategory_ReturnsTrue() = async {
        let attr = DependsOnModulesInCategoryAttribute("infrastructure")
        let context = MockDependencyContext().WithCategory(typeof<CategoryAttributeTestModule>, "infrastructure")
        let result = attr.ShouldDependOn(typeof<CategoryAttributeTestModule>, context)
        do! check(Assert.That(result).IsTrue())
    }

    [<Test>]
    member _.ShouldDependOn_ModuleInDifferentCategory_ReturnsFalse() = async {
        let attr = DependsOnModulesInCategoryAttribute("infrastructure")
        let context = MockDependencyContext().WithCategory(typeof<CategoryAttributeTestModule>, "build")
        let result = attr.ShouldDependOn(typeof<CategoryAttributeTestModule>, context)
        do! check(Assert.That(result).IsFalse())
    }

    [<Test>]
    member _.ShouldDependOn_CaseInsensitive_ReturnsTrue() = async {
        let attr = DependsOnModulesInCategoryAttribute("INFRASTRUCTURE")
        let context = MockDependencyContext().WithCategory(typeof<CategoryAttributeTestModule>, "infrastructure")
        let result = attr.ShouldDependOn(typeof<CategoryAttributeTestModule>, context)
        do! check(Assert.That(result).IsTrue())
    }

    [<Test>]
    member _.ShouldDependOn_ModuleHasNoCategory_ReturnsFalse() = async {
        let attr = DependsOnModulesInCategoryAttribute("infrastructure")
        let context = MockDependencyContext()
        let result = attr.ShouldDependOn(typeof<CategoryAttributeTestModule>, context)
        do! check(Assert.That(result).IsFalse())
    }

    [<Test>]
    member _.ShouldDependOn_CategoryMatchesExactly_ReturnsTrue() = async {
        let attr = DependsOnModulesInCategoryAttribute("build-pipeline")
        let context = MockDependencyContext().WithCategory(typeof<CategoryAttributeTestModule>, "build-pipeline")
        let result = attr.ShouldDependOn(typeof<CategoryAttributeTestModule>, context)
        do! check(Assert.That(result).IsTrue())
    }

    [<Test>]
    member _.Constructor_WithNullCategory_ThrowsArgumentException() = async {
        let mutable threw = false

        try
            DependsOnModulesInCategoryAttribute(null) |> ignore
        with :? ArgumentException ->
            threw <- true

        do! check(Assert.That(threw).IsTrue())
    }

    [<Test>]
    member _.Constructor_WithEmptyCategory_ThrowsArgumentException() = async {
        let mutable threw = false

        try
            DependsOnModulesInCategoryAttribute(String.Empty) |> ignore
        with :? ArgumentException ->
            threw <- true

        do! check(Assert.That(threw).IsTrue())
    }

    [<Test>]
    member _.Constructor_WithWhitespaceCategory_ThrowsArgumentException() = async {
        let mutable threw = false

        try
            DependsOnModulesInCategoryAttribute("   ") |> ignore
        with :? ArgumentException ->
            threw <- true

        do! check(Assert.That(threw).IsTrue())
    }

    [<Test>]
    member _.Category_Property_ReturnsConstructorValue() = async {
        let attr = DependsOnModulesInCategoryAttribute("my-category")
        do! check(StringEqualsAssertionExtensions.IsEqualTo(Assert.That(attr.Category), "my-category"))
    }
